### ISSUE TITLE
WebP image upload preview fails

### ISSUE THREAD

**Issue by dkirkham** [wagtail/wagtail#10947]
### Issue Summary

When using the image upload page in the admin interface, WebP images are not rendered as a preview while the image file is being uploaded, or once the file has completed upload.

### Steps to Reproduce

1. Start a new project with `wagtail start myproject` as per the [Quick Install](https://docs.wagtail.org/en/stable/getting_started/quick_install.html#install-wagtail);
2. Download https://www.gstatic.com/webp/gallery/1.jpg and https://www.gstatic.com/webp/gallery/1.webp into a convenient location (other JPG and WEBP files can be substituted);
3. Browse to the new Wagtail installation, log in and navigate to the _Images_ page, and then the _Add an Image_ page;
4. Upload 1.jpg, either via drag-and-drop or selecting an image file. Observe that the image appears almost immediately at the left with an upload progress bar. Once uploaded the image remains and a form is presented allowing attributes of the image to be adjusted;
5. Upload 1.webp. Observe that the image does not appear at the left, but the progress bar does. Once uploaded the image still does not appear but the form is presented allowing attributes of the image to be adjusted;
6. Observe that nothing is logged in the browser console;
7. Return to the image gallery and observe that the images have both been uploaded.

_Note: it will be necessary to use a rate limiter in responsive design mode within the browser dev tools to see the progress bar, as I've done in the demonstration below._

https://github.com/wagtail/wagtail/assets/1977376/b4fd51ac-4a7e-46bc-9157-8bf5f3bd1c42

Any other relevant information. For example, why do you consider this a bug and what did you expect to happen instead?

During the upload of the WebP image, the image preview should appear on the left, like it does for the JPG image.

I have confirmed that this issue can be reproduced as described on a fresh Wagtail project: **yes**

Note: I have also observed the same issue when uploading _some_ SVG images. I will add to this issue if/when I find example images I can share.

### Technical details

- Python version: 3.10.4
- Django version: 4.2.5.
- Wagtail version: 5.1.2.
- Browser versions:
  -  Firefox 114.0.2, Chrome 117.0.5938.92, Safari 16.5 (18615.2.9.11.4), all on Mac with Apple M1, Ventura 13.4
  -  Firefox 112.0.2, Chrome 117.0.5938.92, Safari 16.2 (18614.3.7.1.5), all on Mac with x86, Ventura 13.1
  -  _Sorry, I don't have ready access to a Windows machine anymore._

**Comment by rohitsrma** [comment]
I can confirm that I've been able to recreate this issue with WebP and AVIF image formats. When using the image upload page in the admin interface, WebP images are not rendered as a preview while the image file is being uploaded. This behavior does not occur when uploading other image formats such as JPG.

**Screenshots:**

![screenshot-127 0 0 1_8000-2023 09 28-18_22_10](https://github.com/wagtail/wagtail/assets/111359305/f314ff92-1ef3-46fb-ad82-634c7996027c)

![screenshot-127 0 0 1_8000-2023 09 28-18_30_21](https://github.com/wagtail/wagtail/assets/111359305/8c7f32aa-f08e-4530-b6bc-55b3a7ee916b)

I'm willing to investigate this issue or making necessary modifications if someone can guide me on where to look in the codebase or provide insights into the problem's root cause. If there are specific files or components related to this issue, please let me know, and I'll do my best to help address it. Thank you!

**Comment by dkirkham** [comment]
I agree with @zerolab, this is an issue in the javascript domain, as the preview image is meant to be rendered before the file is uploaded, let alone downloaded. In the working (JPG) scenario, there is only ever one POST with the file upload, which returns an HTML snippet containing the form, which when inserted into the DOM allows the admin to change some image parameters or even delete the image.

Again, in the working scenario, the preview image is rendered via a <canvas> element, not an <img> element. In the failure scenario, that <canvas> is generally not there - or if it is, it is 0x0px in size.

**Comment by rohitsrma** [comment]
Alright, it seems that the issue could be connected either to these specific files, [jquery.fileupload.js](https://github.com/wagtail/wagtail/blob/a1bd357092bc7107c3d41477536f471a1f3f307d/wagtail/admin/static_src/wagtailadmin/js/vendor/jquery.fileupload.js) and [jquery.fileupload-process.js ](https://github.com/wagtail/wagtail/blob/a1bd357092bc7107c3d41477536f471a1f3f307d/wagtail/admin/static_src/wagtailadmin/js/vendor/jquery.fileupload-process.js)or potentially with the package itself. My confusion arises because when I tried to upload a webp image on the 'https://github.com/blueimp/jQuery-File-Upload' demo project, it triggered an error message saying, 'File type not allowed.' However, when I uploaded a PNG image, it worked fine and displayed a preview within a `<canvas>` tag.

![image](https://github.com/wagtail/wagtail/assets/111359305/70e1da98-2fcb-444a-ab7a-b56410859ccc)

I'm a bit confused by this situation, and I would greatly appreciate any suggestions or guidance anyone might have to offer.

**Comment by zerolab** [comment]
@rohitsrma https://blueimp.github.io/jQuery-File-Upload/ is specifically limited to JPG, GIF, PNG. You would have to tweak that locally to allow WebP

**Comment by dkirkham** [comment]
There appear to be a multitude of issues here.

Firstly the code at https://github.com/blueimp/jQuery-File-Upload has recently been placed into archive mode, so presumably it won't get any maintenance fixes, let alone new features like WebP and AVIF support. It is also based on jQuery which is reaching the end of its useful life, and as [RFC78](https://github.com/wagtail/rfcs/pull/78) states, there is a desire to remove Wagtail's dependence on it.

Secondly, I found this [Stack Overflow answer](https://stackoverflow.com/a/55896125/12578589) which suggests strongly not to use `<canvas>` as a means to perform image conversion. I would concur, the results when previewing some images, and SVGs in particular, are very poor.

Finally, supporting the ideas that both jQuery and `<canvas>` aren't necessary is [this code](https://developer.mozilla.org/en-US/docs/Web/API/FileReader/readAsDataURL) on developer.mozilla.org, which quite successfully renders a local image file image in a browser, with high fidelity (at least in desktop Firefox, Chrome and Safari). The UI there is bare-bones, and of course it has no actual upload capability, but I'm sure some decent HTML and Stimulus code could use the principle here and ultimately replicate all of the current Image upload functionality.

I recognise this is a big change, so I'd leave it to the core team to determine the priority for this within the overall effort to move to Stimulus. It does look like an ideal project for someone who wants to get into a reasonably meaty front-end problem.

A question perhaps worth asking is, how difficult would it be to enhance the current jQuery-File-Upload implementation as a stopgap, versus a complete refactor? Certainly the bug I reported is a deficiency, but it certainly isn't a show-stopper - the images are uploaded to the gallery and correctly appear in choosers, etc.

**Comment by lb-** [comment]
There is a point where we will want to migrate the upload/preview code to Stimulus, maybe it's worth considering this also. 

I had done some research last year on what's in the ecosystem but never got too deep into it.

I'll add a reminder for this to be discussed at the next core team meeting.

**Comment by laymonage** [comment]
Note that this also affects HEIC image upload that we introduced in 6.3.

**Comment by laymonage** [comment]
Good point, thanks. Yep, I don't mean we should have it working on all browsers. I just thought it'd be good to maybe have a better UI fallback, like a placeholder icon. Having just the filename on the left side at that position looks a bit off. And if the browser does support it, maybe make it work if it's not too much effort.

**Comment by thibaudcolas** [comment]
I see our implementation even already comes with a placeholder image, although it is only visible for a few milliseconds until the upload progress starts. I think until we have a better image uploader displaying this placeholder in this scenario is a fine solution.

Current visual (only visible for a few milliseconds):

<img width="837" height="233" alt="Image" src="https://github.com/user-attachments/assets/913d02e9-50c4-473d-801b-133a6d3a564b" />

Code:

https://github.com/wagtail/wagtail/blob/530182ef00672c22f7a070e1bce1b72e95545d0f/wagtail/images/templates/wagtailimages/multiple/add.html#L65-L67

---

Here is the code to change, which hides the placeholder icon:

https://github.com/wagtail/wagtail/blob/530182ef00672c22f7a070e1bce1b72e95545d0f/wagtail/images/static_src/wagtailimages/js/add-multiple.js#L29-L32

I expect we could hide it differently (`display: none` / `hidden` attribute?), then show it again in the `done` callback in the scenario(s) where we detect the preview isn’t displaying:

https://github.com/wagtail/wagtail/blob/530182ef00672c22f7a070e1bce1b72e95545d0f/wagtail/images/static_src/wagtailimages/js/add-multiple.js#L134

**Comment by xevrion** [comment]
Hi! I’ve set up Wagtail locally and was able to reproduce this issue with WebP uploads.

I’m planning to work on a small fix to keep or restore the placeholder icon when the preview isn’t generated (as discussed above), without changing the upload flow itself.

I’ll open a PR once I have something working — please let me know if there’s anything I should watch out for.
